File: mme-tagging-service/internal/security/middleware.go
package security

import (
	"fmt"
	"log"
	"strings"
	"sync"
	"time"

	"github.com/gofiber/fiber/v2"
)

// SecurityConfig holds security configuration
type SecurityConfig struct {
	RateLimitEnabled    bool
	ThreatDetection     bool
	AuditLogging        bool
	SecurityHeaders     bool
	InputValidation     bool
	MaxRequestSize      int64
	RateLimitPerMinute  int
}

// DefaultSecurityConfig returns default security configuration
func DefaultSecurityConfig() SecurityConfig {
	return SecurityConfig{
		RateLimitEnabled:    true,
		ThreatDetection:     true,
		AuditLogging:        true,
		SecurityHeaders:     true,
		InputValidation:     true,
		MaxRequestSize:      10 * 1024 * 1024, // 10MB
		RateLimitPerMinute:  100,
	}
}

// SecurityMiddleware provides comprehensive security for MME Tagging Service
type SecurityMiddleware struct {
	config      SecurityConfig
	rateLimiter *RateLimiter
	auditor     *SecurityAuditor
	detector    *ThreatDetector
}

// NewSecurityMiddleware creates a new security middleware
func NewSecurityMiddleware(config SecurityConfig) *SecurityMiddleware {
	return &SecurityMiddleware{
		config:      config,
		rateLimiter: NewRateLimiter(config.RateLimitPerMinute),
		auditor:     NewSecurityAuditor(),
		detector:    NewThreatDetector(),
	}
}

// SecurityHeaders middleware adds security headers
func (sm *SecurityMiddleware) SecurityHeaders() fiber.Handler {
	if !sm.config.SecurityHeaders {
		return func(c *fiber.Ctx) error {
			return c.Next()
		}
	}

	return func(c *fiber.Ctx) error {
		// Security headers
		c.Set("X-Content-Type-Options", "nosniff")
		c.Set("X-Frame-Options", "DENY")
		c.Set("X-XSS-Protection", "1; mode=block")
		c.Set("Referrer-Policy", "strict-origin-when-cross-origin")
		c.Set("Content-Security-Policy", "default-src 'self'")
		c.Set("Strict-Transport-Security", "max-age=31536000; includeSubDomains")

		return c.Next()
	}
}

// RateLimit middleware implements rate limiting
func (sm *SecurityMiddleware) RateLimit() fiber.Handler {
	if !sm.config.RateLimitEnabled {
		return func(c *fiber.Ctx) error {
			return c.Next()
		}
	}

	return func(c *fiber.Ctx) error {
		userID := c.Get("X-User-ID")
		clientIP := c.IP()
		
		if !sm.rateLimiter.IsAllowed(userID, clientIP) {
			if sm.config.AuditLogging {
				sm.auditor.LogSecurityEvent("RATE_LIMIT_VIOLATION", userID, clientIP, 
					fmt.Sprintf("Rate limit exceeded for endpoint %s", c.Path()))
			}

			return c.Status(fiber.StatusTooManyRequests).JSON(fiber.Map{
				"error": "Rate limit exceeded",
				"code":  "RATE_LIMIT_EXCEEDED",
				"retry_after": 60,
			})
		}

		return c.Next()
	}
}

// ThreatDetection middleware detects malicious requests
func (sm *SecurityMiddleware) ThreatDetection() fiber.Handler {
	if !sm.config.ThreatDetection {
		return func(c *fiber.Ctx) error {
			return c.Next()
		}
	}

	return func(c *fiber.Ctx) error {
		userID := c.Get("X-User-ID")
		clientIP := c.IP()
		
		// Check for threats
		if threat := sm.detector.AnalyzeRequest(c); threat != nil {
			if sm.config.AuditLogging {
				sm.auditor.LogSecurityEvent("THREAT_DETECTED", userID, clientIP, 
					fmt.Sprintf("Threat: %s - %s", threat.Type, threat.Description))
			}

			return c.Status(fiber.StatusForbidden).JSON(fiber.Map{
				"error": "Request blocked by security policy",
				"code":  "SECURITY_VIOLATION",
			})
		}

		return c.Next()
	}
}

// InputValidation middleware validates request input
func (sm *SecurityMiddleware) InputValidation() fiber.Handler {
	if !sm.config.InputValidation {
		return func(c *fiber.Ctx) error {
			return c.Next()
		}
	}

	return func(c *fiber.Ctx) error {
		userID := c.Get("X-User-ID")
		clientIP := c.IP()

		// Check request size
		if c.Context().Request.Header.ContentLength() > int(sm.config.MaxRequestSize) {
			if sm.config.AuditLogging {
				sm.auditor.LogSecurityEvent("OVERSIZED_REQUEST", userID, clientIP, 
					fmt.Sprintf("Request size: %d bytes", c.Context().Request.Header.ContentLength()))
			}

			return c.Status(fiber.StatusRequestEntityTooLarge).JSON(fiber.Map{
				"error": "Request too large",
				"code":  "REQUEST_TOO_LARGE",
			})
		}

		// Check for path traversal
		path := c.Path()
		if strings.Contains(path, "..") || strings.Contains(path, "~") {
			if sm.config.AuditLogging {
				sm.auditor.LogSecurityEvent("PATH_TRAVERSAL_ATTEMPT", userID, clientIP, 
					fmt.Sprintf("Suspicious path: %s", path))
			}

			return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
				"error": "Invalid request path",
				"code":  "INVALID_PATH",
			})
		}

		return c.Next()
	}
}

// AuthenticationMiddleware validates user authentication
func (sm *SecurityMiddleware) AuthenticationMiddleware() fiber.Handler {
	return func(c *fiber.Ctx) error {
		// JWT authentication is handled by Traefik gateway
		// Extract user information from gateway headers
		userID := c.Get("X-User-ID")
		userRoles := c.Get("X-User-Roles")

		// Set user context for downstream handlers
		if userID != "" {
			c.Locals("userID", userID)
		}
		if userRoles != "" {
			c.Locals("userRoles", userRoles)
		}

		return c.Next()
	}
}

// RoleBasedAccessControl creates role-based access control middleware
func (sm *SecurityMiddleware) RoleBasedAccessControl(requiredRole string) fiber.Handler {
	return func(c *fiber.Ctx) error {
		userRoles := c.Get("X-User-Roles")
		userID := c.Get("X-User-ID")
		
		if !hasRequiredRole(userRoles, requiredRole) {
			if sm.config.AuditLogging {
				sm.auditor.LogSecurityEvent("UNAUTHORIZED_ACCESS", userID, c.IP(), 
					fmt.Sprintf("Access denied to %s - required role: %s", c.Path(), requiredRole))
			}

			return c.Status(fiber.StatusForbidden).JSON(fiber.Map{
				"error": "Access denied - insufficient privileges",
				"code":  "INSUFFICIENT_PRIVILEGES",
				"required_role": requiredRole,
			})
		}

		return c.Next()
	}
}

// hasRequiredRole checks if user has the required role
func hasRequiredRole(userRoles, requiredRole string) bool {
	if userRoles == "" {
		return false
	}

	// Check for admin role (has all permissions)
	if strings.Contains(userRoles, "ADMIN") {
		return true
	}

	// Check for specific role
	return strings.Contains(userRoles, requiredRole)
}

// RateLimiter implements in-memory rate limiting
type RateLimiter struct {
	requests map[string]*RateLimitEntry
	limit    int
	mutex    sync.RWMutex
}

type RateLimitEntry struct {
	Count     int
	ResetTime time.Time
}

func NewRateLimiter(limit int) *RateLimiter {
	rl := &RateLimiter{
		requests: make(map[string]*RateLimitEntry),
		limit:    limit,
	}
	
	// Start cleanup goroutine
	go rl.cleanup()
	
	return rl
}

func (rl *RateLimiter) IsAllowed(userID, clientIP string) bool {
	rl.mutex.Lock()
	defer rl.mutex.Unlock()

	key := fmt.Sprintf("%s:%s", userID, clientIP)
	now := time.Now()

	entry, exists := rl.requests[key]
	if !exists {
		rl.requests[key] = &RateLimitEntry{
			Count:     1,
			ResetTime: now.Add(time.Minute),
		}
		return true
	}

	// Reset if time window expired
	if now.After(entry.ResetTime) {
		entry.Count = 1
		entry.ResetTime = now.Add(time.Minute)
		return true
	}

	if entry.Count >= rl.limit {
		return false
	}

	entry.Count++
	return true
}

func (rl *RateLimiter) cleanup() {
	ticker := time.NewTicker(5 * time.Minute)
	defer ticker.Stop()

	for range ticker.C {
		rl.mutex.Lock()
		now := time.Now()
		for key, entry := range rl.requests {
			if now.After(entry.ResetTime.Add(5 * time.Minute)) {
				delete(rl.requests, key)
			}
		}
		rl.mutex.Unlock()
	}
}

// SecurityAuditor handles security event logging
type SecurityAuditor struct{}

func NewSecurityAuditor() *SecurityAuditor {
	return &SecurityAuditor{}
}

func (sa *SecurityAuditor) LogSecurityEvent(eventType, userID, clientIP, description string) {
	log.Printf("[SECURITY EVENT] Type: %s, User: %s, IP: %s, Description: %s, Service: mme-tagging-service, Timestamp: %s",
		eventType, userID, clientIP, description, time.Now().Format(time.RFC3339))
}

// ThreatDetector analyzes requests for security threats
type ThreatDetector struct{}

type Threat struct {
	Type        string
	Severity    string
	Description string
}

func NewThreatDetector() *ThreatDetector {
	return &ThreatDetector{}
}

func (td *ThreatDetector) AnalyzeRequest(c *fiber.Ctx) *Threat {
	// Check query parameters for SQL injection
	queryString := string(c.Context().QueryArgs().QueryString())
	if td.containsSQLInjection(queryString) {
		return &Threat{
			Type:        "SQL_INJECTION",
			Severity:    "HIGH",
			Description: "SQL injection attempt detected in query parameters",
		}
	}

	// Check for XSS attempts
	if td.containsXSS(queryString) {
		return &Threat{
			Type:        "XSS_ATTEMPT",
			Severity:    "MEDIUM",
			Description: "Cross-site scripting attempt detected",
		}
	}

	// Check user agent for suspicious bots
	userAgent := c.Get("User-Agent")
	if td.isSuspiciousUserAgent(userAgent) {
		return &Threat{
			Type:        "SUSPICIOUS_BOT",
			Severity:    "LOW",
			Description: "Suspicious bot or scanner detected",
		}
	}

	return nil
}

func (td *ThreatDetector) containsSQLInjection(input string) bool {
	sqlKeywords := []string{
		"union", "select", "insert", "update", "delete", "drop", "create", "alter",
		"exec", "execute", "sp_", "xp_", "script", "javascript", "vbscript",
	}

	inputLower := strings.ToLower(input)
	for _, keyword := range sqlKeywords {
		if strings.Contains(inputLower, keyword) {
			return true
		}
	}
	return false
}

func (td *ThreatDetector) containsXSS(input string) bool {
	xssPatterns := []string{
		"<script", "</script>", "javascript:", "onload=", "onerror=", "onclick=",
		"alert(", "document.cookie", "window.location", "eval(",
	}

	inputLower := strings.ToLower(input)
	for _, pattern := range xssPatterns {
		if strings.Contains(inputLower, pattern) {
			return true
		}
	}
	return false
}

func (td *ThreatDetector) isSuspiciousUserAgent(userAgent string) bool {
	suspiciousAgents := []string{
		"nmap", "sqlmap", "nikto", "burp", "metasploit", "scanner", "crawler",
		"bot", "spider", "scraper", "harvest", "extract",
	}

	userAgentLower := strings.ToLower(userAgent)
	for _, agent := range suspiciousAgents {
		if strings.Contains(userAgentLower, agent) {
			return true
		}
	}
	return false
}

File: mme-tagmaker-service/app/security/middleware.py
"""
Security middleware for MME Tagmaker Service (FastAPI)
Provides comprehensive security features including rate limiting, threat detection,
audit logging, and role-based access control.
"""

import time
import asyncio
import logging
from collections import defaultdict, Counter
from typing import Dict, Optional, Set, List
from datetime import datetime, timedelta
from fastapi import Request, HTTPException, status
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse
import re
import json

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class SecurityConfig:
    """Security configuration for MME Tagmaker Service"""
    
    def __init__(self):
        self.rate_limit_enabled = True
        self.threat_detection = True
        self.audit_logging = True
        self.security_headers = True
        self.input_validation = True
        self.max_request_size = 10 * 1024 * 1024  # 10MB
        self.rate_limit_per_minute = 100
        self.rate_limit_burst = 20

class RateLimiter:
    """Advanced rate limiter with sliding window and burst protection"""
    
    def __init__(self, config: SecurityConfig):
        self.config = config
        self.requests: Dict[str, List[float]] = defaultdict(list)
        self.blocked: Dict[str, float] = {}
        self.cleanup_interval = 300  # 5 minutes
        
        # Start cleanup task
        asyncio.create_task(self._cleanup_task())
    
    async def _cleanup_task(self):
        """Periodically clean up old entries"""
        while True:
            await asyncio.sleep(self.cleanup_interval)
            await self._cleanup()
    
    async def _cleanup(self):
        """Remove expired entries"""
        current_time = time.time()
        window_start = current_time - 60  # 1 minute window
        
        # Clean up request history
        for key in list(self.requests.keys()):
            self.requests[key] = [t for t in self.requests[key] if t > window_start]
            if not self.requests[key]:
                del self.requests[key]
        
        # Clean up blocked IPs
        for key in list(self.blocked.keys()):
            if current_time - self.blocked[key] > 300:  # 5 minute block
                del self.blocked[key]
    
    def is_allowed(self, user_id: str, client_ip: str, endpoint: str) -> tuple[bool, int]:
        """Check if request is allowed based on rate limits"""
        current_time = time.time()
        key = f"{user_id}:{client_ip}:{endpoint}"
        
        # Check if currently blocked
        if key in self.blocked:
            if current_time - self.blocked[key] < 300:  # 5 minute block
                return False, 300
            else:
                del self.blocked[key]
        
        # Clean old requests (sliding window)
        window_start = current_time - 60
        self.requests[key] = [t for t in self.requests[key] if t > window_start]
        
        # Check rate limit
        request_count = len(self.requests[key])
        limit = self._get_endpoint_limit(endpoint)
        
        if request_count >= limit:
            self.blocked[key] = current_time
            return False, 300
        
        # Add current request
        self.requests[key].append(current_time)
        return True, 0
    
    def _get_endpoint_limit(self, endpoint: str) -> int:
        """Get rate limit based on endpoint"""
        if "/extract" in endpoint or "/tag" in endpoint:
            return 50  # AI endpoints have lower limits
        elif "/health" in endpoint or "/metrics" in endpoint:
            return 200  # Health checks have higher limits
        return self.config.rate_limit_per_minute

class ThreatDetector:
    """Advanced threat detection system"""
    
    def __init__(self):
        self.sql_patterns = [
            r"union.*select", r"insert.*into", r"delete.*from", r"update.*set",
            r"drop.*table", r"create.*table", r"alter.*table", r"exec.*sp_",
            r"xp_cmdshell", r"sp_executesql", r"openrowset", r"opendatasource"
        ]
        
        self.xss_patterns = [
            r"<script.*?>", r"javascript:", r"onload\s*=", r"onerror\s*=",
            r"onclick\s*=", r"alert\s*\(", r"document\.cookie", r"window\.location",
            r"eval\s*\(", r"expression\s*\(", r"vbscript:", r"data:text/html"
        ]
        
        self.injection_patterns = [
            r"\.\./", r"\.\.\\", r"~.*?/", r"/etc/passwd", r"/etc/shadow",
            r"cmd\.exe", r"powershell", r"/bin/sh", r"/bin/bash"
        ]
        
        self.bot_patterns = [
            r"nmap", r"sqlmap", r"nikto", r"burp", r"metasploit", r"scanner",
            r"crawler", r"spider", r"scraper", r"harvest", r"extract", r"bot"
        ]
    
    def analyze_request(self, request: Request) -> Optional[Dict]:
        """Analyze request for security threats"""
        threats = []
        
        # Analyze URL and query parameters
        url_str = str(request.url).lower()
        query_str = str(request.url.query).lower() if request.url.query else ""
        
        # Check for SQL injection
        if self._check_patterns(url_str + query_str, self.sql_patterns):
            threats.append({
                "type": "SQL_INJECTION",
                "severity": "HIGH",
                "description": "SQL injection attempt detected in URL/query"
            })
        
        # Check for XSS
        if self._check_patterns(url_str + query_str, self.xss_patterns):
            threats.append({
                "type": "XSS_ATTEMPT",
                "severity": "MEDIUM",
                "description": "Cross-site scripting attempt detected"
            })
        
        # Check for path traversal
        if self._check_patterns(url_str, self.injection_patterns):
            threats.append({
                "type": "PATH_TRAVERSAL",
                "severity": "HIGH",
                "description": "Path traversal attempt detected"
            })
        
        # Check user agent
        user_agent = request.headers.get("user-agent", "").lower()
        if self._check_patterns(user_agent, self.bot_patterns):
            threats.append({
                "type": "SUSPICIOUS_BOT",
                "severity": "LOW",
                "description": "Suspicious bot or scanner detected"
            })
        
        return threats[0] if threats else None
    
    def _check_patterns(self, text: str, patterns: List[str]) -> bool:
        """Check text against regex patterns"""
        for pattern in patterns:
            if re.search(pattern, text, re.IGNORECASE):
                return True
        return False

class SecurityAuditor:
    """Security event audit logger"""
    
    def __init__(self):
        self.security_logger = logging.getLogger("security_audit")
        handler = logging.StreamHandler()
        formatter = logging.Formatter(
            '%(asctime)s - [SECURITY EVENT] - %(levelname)s - %(message)s'
        )
        handler.setFormatter(formatter)
        self.security_logger.addHandler(handler)
        self.security_logger.setLevel(logging.WARNING)
    
    def log_security_event(self, event_type: str, user_id: str, client_ip: str, 
                          description: str, endpoint: str = ""):
        """Log security event with structured data"""
        event_data = {
            "event_type": event_type,
            "user_id": user_id,
            "client_ip": client_ip,
            "description": description,
            "endpoint": endpoint,
            "service": "mme-tagmaker-service",
            "timestamp": datetime.utcnow().isoformat()
        }
        
        self.security_logger.warning(json.dumps(event_data))

class SecurityMetrics:
    """Security metrics collector"""
    
    def __init__(self):
        self.rate_limit_violations = Counter()
        self.threat_detections = Counter()
        self.unauthorized_access = Counter()
        self.threats_by_type = Counter()
        self.violations_by_endpoint = Counter()
        self.last_updated = datetime.utcnow()
    
    def record_rate_limit_violation(self, endpoint: str, user_id: str):
        """Record rate limit violation"""
        self.rate_limit_violations[f"{user_id}:{endpoint}"] += 1
        self.violations_by_endpoint[endpoint] += 1
        self.last_updated = datetime.utcnow()
    
    def record_threat_detection(self, threat_type: str, user_id: str):
        """Record threat detection"""
        self.threat_detections[f"{user_id}:{threat_type}"] += 1
        self.threats_by_type[threat_type] += 1
        self.last_updated = datetime.utcnow()
    
    def record_unauthorized_access(self, user_id: str):
        """Record unauthorized access attempt"""
        self.unauthorized_access[user_id] += 1
        self.last_updated = datetime.utcnow()
    
    def get_metrics(self) -> Dict:
        """Get current security metrics"""
        return {
            "rate_limit_violations": sum(self.rate_limit_violations.values()),
            "threat_detections": sum(self.threat_detections.values()),
            "unauthorized_access": sum(self.unauthorized_access.values()),
            "threats_by_type": dict(self.threats_by_type),
            "violations_by_endpoint": dict(self.violations_by_endpoint),
            "last_updated": self.last_updated.isoformat()
        }

class SecurityMiddleware(BaseHTTPMiddleware):
    """Comprehensive security middleware for FastAPI"""
    
    def __init__(self, app, config: SecurityConfig = None):
        super().__init__(app)
        self.config = config or SecurityConfig()
        self.rate_limiter = RateLimiter(self.config)
        self.threat_detector = ThreatDetector()
        self.auditor = SecurityAuditor()
        self.metrics = SecurityMetrics()
    
    async def dispatch(self, request: Request, call_next):
        """Main security middleware dispatch"""
        start_time = time.time()
        
        # Extract user information from gateway headers
        user_id = request.headers.get("x-user-id", "anonymous")
        user_roles = request.headers.get("x-user-roles", "")
        client_ip = self._get_client_ip(request)
        endpoint = request.url.path
        
        try:
            # Security headers
            if self.config.security_headers:
                response = await self._add_security_headers(request, call_next)
            else:
                response = await call_next(request)
            
            # Input validation
            if self.config.input_validation and not await self._validate_input(request, user_id, client_ip):
                return JSONResponse(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    content={"error": "Invalid request", "code": "INVALID_REQUEST"}
                )
            
            # Rate limiting
            if self.config.rate_limit_enabled:
                allowed, retry_after = self.rate_limiter.is_allowed(user_id, client_ip, endpoint)
                if not allowed:
                    self.auditor.log_security_event(
                        "RATE_LIMIT_VIOLATION", user_id, client_ip,
                        f"Rate limit exceeded for endpoint {endpoint}", endpoint
                    )
                    self.metrics.record_rate_limit_violation(endpoint, user_id)
                    
                    return JSONResponse(
                        status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                        content={
                            "error": "Rate limit exceeded",
                            "code": "RATE_LIMIT_EXCEEDED",
                            "retry_after": retry_after
                        }
                    )
            
            # Threat detection
            if self.config.threat_detection:
                threat = self.threat_detector.analyze_request(request)
                if threat:
                    self.auditor.log_security_event(
                        "THREAT_DETECTED", user_id, client_ip,
                        f"Threat: {threat['type']} - {threat['description']}", endpoint
                    )
                    self.metrics.record_threat_detection(threat["type"], user_id)
                    
                    return JSONResponse(
                        status_code=status.HTTP_403_FORBIDDEN,
                        content={
                            "error": "Request blocked by security policy",
                            "code": "SECURITY_VIOLATION"
                        }
                    )
            
            # Process time for monitoring
            process_time = time.time() - start_time
            response.headers["X-Process-Time"] = str(process_time)
            
            return response
            
        except Exception as e:
            logger.error(f"Security middleware error: {e}")
            return JSONResponse(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                content={"error": "Internal security error", "code": "SECURITY_ERROR"}
            )
    
    async def _add_security_headers(self, request: Request, call_next):
        """Add security headers to response"""
        response = await call_next(request)
        
        # Security headers
        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["X-XSS-Protection"] = "1; mode=block"
        response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"
        response.headers["Content-Security-Policy"] = "default-src 'self'"
        response.headers["Strict-Transport-Security"] = "max-age=31536000; includeSubDomains"
        
        return response
    
    async def _validate_input(self, request: Request, user_id: str, client_ip: str) -> bool:
        """Validate request input"""
        # Check request size
        content_length = request.headers.get("content-length")
        if content_length and int(content_length) > self.config.max_request_size:
            self.auditor.log_security_event(
                "OVERSIZED_REQUEST", user_id, client_ip,
                f"Request size: {content_length} bytes", request.url.path
            )
            return False
        
        # Check for path traversal
        path = str(request.url.path)
        if ".." in path or "~" in path:
            self.auditor.log_security_event(
                "PATH_TRAVERSAL_ATTEMPT", user_id, client_ip,
                f"Suspicious path: {path}", path
            )
            return False
        
        return True
    
    def _get_client_ip(self, request: Request) -> str:
        """Extract client IP from request headers"""
        # Check for forwarded headers (from reverse proxy)
        forwarded_for = request.headers.get("x-forwarded-for")
        if forwarded_for:
            return forwarded_for.split(",")[0].strip()
        
        real_ip = request.headers.get("x-real-ip")
        if real_ip:
            return real_ip
        
        return request.client.host if request.client else "unknown"

def has_required_role(user_roles: str, required_role: str) -> bool:
    """Check if user has the required role"""
    if not user_roles:
        return False
    
    # Admin role has all permissions
    if "ADMIN" in user_roles:
        return True
    
    # Check for specific role
    return required_role in user_roles

# Global instance for access from handlers
security_middleware_instance = None

def require_role(required_role: str):
    """Decorator to require specific role for endpoint access"""
    def decorator(func):
        async def wrapper(*args, **kwargs):
            # Extract request from args/kwargs
            request = None
            for arg in args:
                if isinstance(arg, Request):
                    request = arg
                    break
            
            if not request:
                raise HTTPException(
                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                    detail="Internal error: Request object not found"
                )
            
            user_roles = request.headers.get("x-user-roles", "")
            if not has_required_role(user_roles, required_role):
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail={
                        "error": "Access denied - insufficient privileges",
                        "code": "INSUFFICIENT_PRIVILEGES",
                        "required_role": required_role
                    }
                )
            
            return await func(*args, **kwargs)
        return wrapper
    return decorator